File: typescript/preliminar/src/app/modules/auth/middlewares/auth.middleware.ts
import { verify } from "jsonwebtoken";

// Types
import { JwtPayload } from "../interfaces/jwt-payload.interface";

// Errors
import { BadRequestError } from "@/app/common/errors/bad-request.error";
import { UnauthorizedError } from "@/app/common/errors/unauthorized.error";

// MESSAGES
import { MESSAGES } from "@/app/common/messages";

// Types
import {
  Request,
  Response,
  NextFunction,
} from "@/app/common/interfaces/http.interfaces";

const SECRET = process.env.JWT_SECRET;

export const authMiddleware = async (
  req: Request,
  res: Response,
  next: NextFunction
) => {
  if (!req?.headers["authorization"]) {
    const error = new UnauthorizedError();
    return res.status(error.statusCode).json(error.toJSON());
  }

  const [type, token] = req?.headers["authorization"]?.split(" ");

  if (!type || type !== "Bearer") {
    const error = new BadRequestError(
      MESSAGES.error.middleware.auth.BadRequest.Unformatted
    );
    return res.status(error.statusCode).json(error.toJSON());
  }

  if (!token) {
    const error = new BadRequestError(
      MESSAGES.error.middleware.auth.BadRequest.Unformatted
    );
    return res.status(error.statusCode).json(error.toJSON());
  }

  const result = verify(token, SECRET) as JwtPayload;

  req.auth = result;

  next();
};

File: go/fiber/internal/modules/auth/middleware/auth_middleware.go
package auth

import (
	"os"
	"strings"
	"tcc/internal/common/enum"
	"tcc/internal/common/errors"
	"tcc/internal/common/messages"
	auth "tcc/internal/modules/auth/interfaces"

	"github.com/gofiber/fiber/v2"
	"github.com/golang-jwt/jwt/v5"
)

var SECRET = []byte(os.Getenv("JWT_SECRET"))

func AuthMiddleware() fiber.Handler {
	return func(c *fiber.Ctx) error {
		authHeader := c.Get("Authorization")
		if authHeader == "" {
			unauthorized := errors.UnauthorizedError(messages.ErrorMessages.Unauthorized)
			return c.Status(unauthorized.StatusCode).JSON(unauthorized)
		}

		parts := strings.Split(authHeader, " ")
		if len(parts) != 2 || parts[0] != "Bearer" {
			badRequest := errors.BadRequestError(messages.ErrorMessages.Middleware.Auth.BadRequest.Unformatted)
			return c.Status(badRequest.StatusCode).JSON(badRequest)
		}

		tokenString := parts[1]
		token, err := jwt.Parse(tokenString, func(token *jwt.Token) (interface{}, error) {
			return SECRET, nil
		})

		if err != nil || !token.Valid {
			unauthorized := errors.UnauthorizedError(messages.ErrorMessages.Unauthorized)
			return c.Status(unauthorized.StatusCode).JSON(unauthorized)
		}

		claims, ok := token.Claims.(jwt.MapClaims)
		if !ok {
			unauthorized := errors.UnauthorizedError(messages.ErrorMessages.Unauthorized)
			return c.Status(unauthorized.StatusCode).JSON(unauthorized)
		}

		c.Locals("auth", auth.JwtPayload{
			CPF:  claims["cpf"].(string),
			Role: enum.RoleEnum(claims["role"].(string)),
			ID: claims["id"].(string),
			Email: claims["email"].(string),
		})
		return c.Next()
	}
}
